File: solved/10116/robot.py
rows, cols, firstcol = [int(x) for x in raw_input().split()]
while(rows != 0):
	grid = [raw_input() for i in range(rows)]
	positions = {}
	curpos = [0, firstcol-1]
	nummoves = 0
	loopcount = 0
	operations = {'N': lambda x: [x[0] - 1, x[1]],
				  'S': lambda x: [x[0] + 1, x[1]],
				  'E': lambda x: [x[0], x[1] + 1],
				  'W': lambda x: [x[0], x[1] - 1]}
	while(curpos[0] >= 0 and curpos[0] < rows and curpos[1] >= 0 and curpos[1] < cols):
		key = "%d%d" % (curpos[0], curpos[1])
		if(positions.has_key(key)):
			val = positions[key]
			loopcount = nummoves - val
			nummoves -= loopcount
			break
		else:
			positions[key] = nummoves
			nummoves += 1
		curpos = operations[grid[curpos[0]][curpos[1]]](curpos)
	if(loopcount != 0):
		print "%d step(s) before a loop of %d step(s)" % (nummoves, loopcount)
	else:
		print "%d step(s) to exit" % nummoves
	rows, cols, firstcol = [int(x) for x in raw_input().split()]

File: solved/10116/robot.cc
#include <iostream>
#include <string>
#include <vector>
#include <map>

using namespace std;

class position {
public:
	int row;
	int col;
	
	position(int r, int c) {
		row = r;
		col = c;
	}
};

int main() {
	int rows = 0;
	int cols = 0;
	int startCol = 0;
	cin >> rows >> cols >> startCol;
	while(rows != 0) {
		string s;
		vector<string> vec;
		map<int, int> moves;
		position curPos(0, 0);
		int numMoves = 0;
		int numLoopMoves = -1;
		
		for(int i = 0; i < rows; i++) {
			cin >> s;
			vec.push_back(s);
		}
		
		curPos.col = startCol - 1;
		moves[10 * curPos.row + curPos.col - 1] = 1;
		
		while((curPos.row != -1) && (curPos.row < rows) && (curPos.col != -1) && (curPos.col) < cols) {
			char c = vec[curPos.row][curPos.col];
			switch(c) {
				case 'N':
				curPos.row--;
				break;
				case 'S':
				curPos.row++;
				break;
				case 'E':
				curPos.col++;
				break;
				case 'W':
				curPos.col--;
				break;
			}
			int index = 10 * curPos.row + curPos.col - 1;
			int n = moves[index];
			if(n != 0) {
				n--;
				numLoopMoves = numMoves + 1 - n;
				numMoves = n;
				break;
			}
			else {
				numMoves++;
				moves[index] = numMoves + 1;
			}
		}
		
		if(numLoopMoves != -1)
			cout << numMoves << " step(s) before a loop of " << numLoopMoves << " step(s)" << endl;
		else
			cout << numMoves << " step(s) to exit" << endl;
		
		cin >> rows >> cols >> startCol;
	}
    return 0;
}
